src/util.rs

```diff
@@ -51,14 +52,14 @@ fn expect_property(obj: &mut JsonObject, name: &'static str) -> Result<JsonValue
 fn expect_owned_array(value: JsonValue) -> Result<Vec<JsonValue>, Error> {
     match value {
         JsonValue::Array(v) => Ok(v),
-        _ => Err(Error::ExpectedArrayValue),
+        _ => Err(Error::ExpectedOwnedArrayValue(value)),
```

Comment:
In the interest of exposing as few error types as is useful, and having helpful errors for the 'unowned case', what would you think about something like:

```
diff --git a/src/errors.rs b/src/errors.rs
index 3d79947..b4df40e 100644
--- a/src/errors.rs
+++ b/src/errors.rs
@@ -38,10 +38,8 @@ pub enum Error {
     ExpectedProperty(String),
     #[error("Expected a floating-point value, but got None")]
     ExpectedF64Value,
-    #[error("Expected an array value, but got None")]
-    ExpectedArrayValue,
-    #[error("Expected an owned array value, but got `{0}`")]
-    ExpectedOwnedArrayValue(Value),
-    #[error("Expected an owned Object, but got `{0}`")]
+    #[error("Expected an Array value, but got {0}")]
+    ExpectedArrayValue(Value),
+    #[error("Expected an Object, but got `{0}`")]
     ExpectedObjectValue(Value),
 }
diff --git a/src/util.rs b/src/util.rs
index 10d8aee..fba79ac 100644
--- a/src/util.rs
+++ b/src/util.rs
@@ -38,7 +38,7 @@ pub fn expect_f64(value: &JsonValue) -> Result<f64, Error> {
 pub fn expect_array(value: &JsonValue) -> Result<&Vec<JsonValue>, Error> {
     match value.as_array() {
         Some(v) => Ok(v),
-        None => Err(Error::ExpectedArrayValue),
+        None => Err(Error::ExpectedArrayValue(value.clone())),  // <- this clone might be the controversial part.
     }
 }
```

Reply:
Do you mean controversial for performance reasons?

Reply:
> Do you mean controversial for performance reasons?

Yep. Personally I think it's probably worth it for the better error ergonomics, just didn't want it to slip by unnoticed.

Reply:
I don't have a good sense of the array sizes we might be cloning (I.e. could it be a polygon outer ring, which _could_ be expensive) so I might try to instrument that tomorrow, but I think this is a good idea.

Reply:
OK, so it looks like we'd be cloning an entire `FeatureCollection` or `GeometryCollection` if this error occurs and we return the entire `Value`, so we're now down the rabbit hole of: "But it's essentially impossible to recover from this anyway, so who cares". I wonder if it would be more useful to communicate that we were expecting an `Array`, but got one of `{Null, Bool, String, Number, Object}` instead.

Reply:
5f62691 collapses the missing Array errors into one because the distinction between `&Array` and `Array` isn't meaningful anyway. Instead, the error indicates which `serde_json::Value` enum member was actually received instead of `Array`.